src/Microsoft.Unity.Analyzers/ProtectedUnityMessage.cs

```diff
@@ -0,0 +1,111 @@
+			id: "UNT0021",
+			title: Strings.ProtectedUnityMessageDiagnosticTitle,
+			messageFormat: Strings.ProtectedUnityMessageDiagnosticMessageFormat,
+			category: DiagnosticCategory.Correctness,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
+			description: Strings.ProtectedUnityMessageDiagnosticDescription);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzeMethodDeclaration, SyntaxKind.MethodDeclaration);
+		}
+
+		private static void AnalyzeMethodDeclaration(SyntaxNodeAnalysisContext context)
+		{
+			var method = (MethodDeclarationSyntax)context.Node;
+			if (method?.Body == null)
+				return;
+
+			if (method.Modifiers.Any(SyntaxKind.ProtectedKeyword))
+				return;
+
+			var classDeclaration = method.FirstAncestorOrSelf<ClassDeclarationSyntax>();
+			if (classDeclaration == null)
+				return;
+
+			var typeSymbol = context.SemanticModel.GetDeclaredSymbol(classDeclaration);
+			var scriptInfo = new ScriptInfo(typeSymbol);
+			if (!scriptInfo.HasMessages)
+				return;
+
+			var symbol = context.SemanticModel.GetDeclaredSymbol(method);
+			if (!scriptInfo.IsMessage(symbol))
+				return;
+
+			context.ReportDiagnostic(Diagnostic.Create(Rule, method.Identifier.GetLocation()));
+		}
+	}
+
+	[ExportCodeFixProvider(LanguageNames.CSharp)]
+	public class ProtectedUnityMessageCodeFix : CodeFixProvider
+	{
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ProtectedUnityMessageAnalyzer.Rule.Id);
+
+		public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			var declaration = root.FindNode(context.Span) as MethodDeclarationSyntax;
+			if (declaration == null)
+				return;
+
+			context.RegisterCodeFix(
+				CodeAction.Create(
+					Strings.ProtectedUnityMessageCodeFixTitle,
+					ct => MakeMessageProtectedAsync(context.Document, declaration, ct),
+					declaration.ToFullString()),
+				context.Diagnostics);
+		}
+
+		private static async Task<Document> MakeMessageProtectedAsync(Document document, MethodDeclarationSyntax declaration, CancellationToken cancellationToken)
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+			var newDeclaration = declaration.WithModifiers(SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.ProtectedKeyword)));
+
+			foreach (var modifier in declaration.Modifiers)
+			{
+				if (!modifier.IsKind(SyntaxKind.PublicKeyword) && !modifier.IsKind(SyntaxKind.PrivateKeyword))
```

Comment:
and what if i'm using an `internal` message ? Hopefully `protected internal` is a valid access modifier but not sure if we want to do that.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/access-modifiers

Reply:
if it's internal, would we still want to make it protected? (remove internal and add protected modifier)

Reply:
Yes let's do that to start.